keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -638,24 +640,185 @@ always dispatched immediately.  Following is how the other requests
+priority level `i`'s seat demand over the just-concluded adjustment
+period.  A priority level's seat demand at any given moment is the sum
+of its occupied seats and the number of seats in the queued requests.
+We also define `EnvelopeSD(i) = AvgSD(i) + StDevSD(i)`.  The
+adjustment logic is driven by a quantity called smoothed seat demand
+(`SmoothSD(i)`), which does an exponential averaging of EnvelopeSD
+values using a coeficient A in the range (0,1) and immediately tracks
+EnvelopeSD when it exceeds SmoothSD.  The rule for updating priority
+level `i`'s SmoothSD at the end of an adjustment period is
+`SmoothSD(i) := max( EnvelopeSD(i), A*SmoothSD(i) + (1-A)*Envelope(i)
+)`.  The command line flag `--seat-demand-history-fraction` with a
+default value of 0.9 configures A.
+
+Adjustment is also done on configuration change, when a priority level
+is introduced or removed or its NominalCL or BorrowableCL changes.  At
+such a time, the current adjustment period comes to an early end and
+the regular adjustment logic runs; the adjustment timer is reset to
+next fire 10 seconds later.  For a newly introduced priority level, we
+set HighSD, AvgSD, and SmoothSD to NominalCL-BorrowableSD/2 and
+StDevSD to zero.
+
+For adjusting the CurrentCL values, each non-exempt priority level `i`
+has a lower bound (`MinCurrentCL(i)`) for the new value.  It is simply
+HighSD clipped by the configured concurrency limits: `MinCurrentCL(i)
+= max( MinCL(i), min( NominalCL(i), HighSD(i) ) )`.
+
+If MinCurrentCL(i) = NominalCL(i) for every non-exempt priority level
+i then there is no wiggle room.  No priority level is willing to lend
+any seats.  The new CurrentCL values must equal the NominalCL values.
+Otherwise there is wiggle room and the adjustment proceeds as follows.
+
+The priority levels would all be fairly happy if we set CurrentCL =
+SmoothSD for each.  We clip that by the lower bound just shown, taking
+`Target(i) = max(SmoothSD(i), MinCurrentCL(i))` as a first-order
+target for each non-exempt priority level `i`.
+
+Sadly, the sum of the Target values --- let's name that TargetSum ---
+is not necessarily equal to ServerCL and the individual Target values
+do not necessarily respect the corresponding MinCurrentCL bound.  If
+we had only the first of those two problems then we could set each
+CurrentCL(i) to FairFrac * Target(i) where FairFrac = ServerCL /
+TargetSum.  This would share the gain or pain in equal proportion
+among the priority levels.  Taking the lower bounds into account means
+finding the one FairFrac value that solves the following conditions,
+for all the non-exempt priority levels `i`, and also makes the
+CurrentCL values sum to ServerCL.  For this step we let the CurrentCL
+values be floating-point numbers, not necessarily integers.
+
+```
+CurrentCL(i) = FairFrac * Target(i)  if FairFrac * Target(i) >= MinCurrentCL(i)
+CurrentCL(i) = MinCurrentCL(i)       if FairFrac * Target(i) <= MinCurrentCL(i)
+```
```

Comment:
At this point, are guaranteed that CurrentCLSum (sum of all CurrentCL(i)) is equal to ServerCL?  I'm not seeing how that is guaranteed.

Reply:
The point is that FairFrac is a variable here that has to be solved for.  Higher FairFrac makes for higher sum of CurrentCL; the problem is to pick the value of FairFrac that makes the CurrentCL sum to ServerCL.

Reply:
I revised the write-up to go through this a little more slowly, hopefully it is clearer now (if a bit longer).